packages/core/utils/lib/async.js

```diff
@@ -20,7 +20,24 @@ function pipeAsync(...methods) {
  */
 const mapAsync = curry(pMap);
 
+/**
+ * Because of how mysql works, making parallel requests can cause deadlocks.
+ * This function will run the requests sequentially if mysql is used. Else,
+ * it will encapsulate them in a Promise.all.
+ *
+ * @type { import('./async').MapAsyncDialects }
+ */
+const mapAsyncDialects = async (array, func, options = {}) => {
+  switch (strapi.db.dialect.client) {
+    case 'mysql':
+      return mapAsync(array, func, { concurrency: 1 });
+    default:
+      return mapAsync(array, func, options);
+  }
+};
```

Comment:
Mmmh what do you think of this utils?
`mapAsyncDialects` sounds generic while it's taking care of a very specific problem.
It feels weird to me that this dialect logic is in `utils` whereas it depends on each situation (we don't use sequential everywhere). It sounds too specific for me to be put in a `utils` like that, maybe if it had a name very explicit like `mapAsyncSequentialOnMySQLAndParallelOtherwise`?
Also, sometimes (like in `localizations`), it's not a `map` that is appropriate but a `forEach`. Would there be another `utils` for this situation?
I would see something simpler by having the dialect logic directly next to the code involved. In `plugins/i18n/server/services/localizations.js` I would see:
```js
  // MySQL/MariaDB can cause deadlocks here if concurrency higher than 1
  await forEachAsync(array, func, { concurrency: strapi.db.dialect.client === 'mysql' ? 1 : Infinity });
```
or using a util to get the dialect to make it look cleaner
```js
  // MySQL/MariaDB can cause deadlocks here if concurrency higher than 1
  await forEachAsync(array, func, { concurrency: isDialectMySQL() ? 1 : Infinity });
```

For me it is dirty fix that should look like a dirty fix (direct and simple fix) that should ultimately be fixed by fixing the core deadlock issue (that still isn't understood properly). If it is handled in a more complex/"clean" way, it way look like legit code that we would not question.
What do you think?

Reply:
I think you are completely on point with this one.
I did not feel this function was in the right place but I did not find any place better.

Doing it like you propose seems ok to me :) About using a forEach, as we are already using pmap and it handles concurrency, do you think we could still use the mapAsync for this matter? Or we should yet implement the forEachAsync for ourselves?

ping @nathan-pichon

Reply:
Cool :)
It would be cleaner to implement a `forEachAsync` I would say. But it could reuse `p-map` for simplicity

Reply:
I agree that this doesn't seem like a util for the whole core part of the app.

For the `forEach` part, it means we don't care about the result of functions. Is it the case? I don't think, or at least, I think we should definitely handle the results of functions to be able to know if it's a success or if we have functional errors to handle, no?

Reply:
I think, in the localization code we do not handle any result for now, https://github.com/strapi/strapi/pull/15643/files#diff-946e46cbda9c114fd47837681eaf582f8c28c85e84f9ef6c1be3b9d0643fc5bf